Question1

In [ ]:
import numpy as np
temperature = np.array([25, 28, 31, 35, 38, 27, 33, 40]) 
corr_temp=temperature+2
print("corrected temperature" ,corr_temp)
fahrenheit = (9/5) * temperature + 32
print("fahrenheit temperature " ,fahrenheit)
readings=temperature[temperature>32]
print("readings are ",readings)
count=np.sum(temperature>32)
print("tottal readings more than 32 are",count)

corrected temperature [27 30 33 37 40 29 35 42]
fahrenheit temperature  [ 77.   82.4  87.8  95.  100.4  80.6  91.4 104. ]
readings are  [35 38 33 40]
tottal readings more than 32 4


1 (e)Vectorization performs operations on the entire array at once, avoiding explicit for loops, making code shorter and faster.

Boolean indexing quickly filters elements based on a condition without looping through each value.

So, both make NumPy data processing more efficient and concise.


Question 2

In [5]:
steps = np.array([ [5000, 6200, 7100], [8000, 7500, 9000], [4500, 5100, 4800], [9000, 8500, 9500]]) 
print("total steps",np.sum(steps))
print("Mean steps", np.mean(steps))
print("Maximum", np.max(steps))
print("Minimum", np.min(steps))
print("Total for each day", np.sum(steps, axis=0))
print("Total for each user", np.sum(steps, axis=1))
position = np.unravel_index(np.argmax(steps), steps.shape)
print("Position of maximum:", position)
print("Maximum steps:", steps[position])

total steps 84200
Mean steps 7016.666666666667
Maximum 9500
Minimum 4500
Total for each day [26500 27300 30400]
Total for each user [18300 24500 14400 27000]
Position of maximum: (np.int64(3), np.int64(2))
Maximum steps: 9500


question3

In [11]:
import numpy as np
original = np.array([1, 2, 3, 4, 5, 6])

subset=original[1:5]
print("subset",subset) 

subset[0] = 999
print("Original", original)
print("Subset", subset)

another = original[1:5].copy()
another[0] = 500
print("Original", original)
print("Copy", another)

arr = np.arange(1, 13)
matrix = arr.reshape(3, 4)
print(matrix)

print("\nFirst row", matrix[0])
print("Last row", matrix[-1])
print("Second column", matrix[:, 1])
print("Rows 1-2, columns 2-3")
print(matrix[1:3, 2:4])

flat = matrix.flatten()
rav = matrix.ravel()
print("\nFlatten:", flat)
print("Ravel:", rav)

rav[0] = 1000
print("\nAfter modifying ravel")
print("Ravel", rav)
print("Original matrix")
print(matrix)

flat[1] = 2000
print("\nAfter modifying flatten")
print("Flatten", flat)
print("Original matrix")
print(matrix)

print("\nShape", matrix.shape)
print("Dimensions", matrix.ndim)
print("Size", matrix.size)
print("Data type", matrix.dtype)

subset [2 3 4 5]
Original [  1 999   3   4   5   6]
Subset [999   3   4   5]
Original [  1 999   3   4   5   6]
Copy [500   3   4   5]
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

First row [1 2 3 4]
Last row [ 9 10 11 12]
Second column [ 2  6 10]
Rows 1-2, columns 2-3
[[ 7  8]
 [11 12]]

Flatten: [ 1  2  3  4  5  6  7  8  9 10 11 12]
Ravel: [ 1  2  3  4  5  6  7  8  9 10 11 12]

After modifying ravel
Ravel [1000    2    3    4    5    6    7    8    9   10   11   12]
Original matrix
[[1000    2    3    4]
 [   5    6    7    8]
 [   9   10   11   12]]

After modifying flatten
Flatten [   1 2000    3    4    5    6    7    8    9   10   11   12]
Original matrix
[[1000    2    3    4]
 [   5    6    7    8]
 [   9   10   11   12]]

Shape (3, 4)
Dimensions 2
Size 12
Data type int64


Question 4

In [20]:
X = np.array([ [6, 70, 3], [5, 50, 6], [8, 80, 2], [4, 30, 8] ]) 
y = np.array([40, 65, 30, 85])

print("Shape", X.shape)
print("Dimensions", X.ndim)

print("transpose",X.T)
print("Transpose swaps rows and columns.After transpose, X.T is 3 × 4,each row of X.T contains all values of one feature across the 4 users. ")

XtX = X.T @ X
print("matrix product",XtX)

inverse = np.linalg.inv(XtX)
print("matrix inverse",inverse)

beta = np.linalg.inv(X.T @ X) @ X.T @ y
print("equation is",beta)

print("\nCoefficients:")
print("Sleep:", beta[0])
print("Activity:", beta[1])
print("Stress:", beta[2])

new_user = np.array([5, 40, 7])
prediction = new_user @ beta
print("Predicted assistance score:", prediction)

Shape (4, 3)
Dimensions 2
transpose [[ 6  5  8  4]
 [70 50 80 30]
 [ 3  6  2  8]]
Transpose swaps rows and columns.After transpose, X.T is 3 × 4,each row of X.T contains all values of one feature across the 4 users. 
matrix product [[  141  1430    96]
 [ 1430 14700   910]
 [   96   910   113]]
matrix inverse [[ 1.10521428 -0.09848746 -0.14581398]
 [-0.09848746  0.00891203  0.01190129]
 [-0.14581398  0.01190129  0.0368847 ]]
equation is [-0.15921454  0.14083853 10.06302242]

Coefficients:
Sleep: -0.15921454159483517
Activity: 0.1408385299190522
Stress: 10.063022422714608
Predicted assistance score: 75.27862544779018


question4 f:
The coefficients show the effect of each feature on the assistance score while keeping the other features constant.

Sleep: -0.159 → slight decrease in score
Activity: +0.141 → slight increase in score
Stress: +10.063 → increase in score
